### PR TITLE
Add block & timestamp conditions to Signer

### PR DESCRIPTION
**Pull request by jacogr**
- Allows the editing of min transmission block or minimum transaction timestamp from Signer, Send Transaction & Execute Contract (anywhere where the GasPriceEditor component was available)
- Depends on Parity support for `condition: { block: <some-blockNumber> }` & `condition: { time: <ms-since-1970> }` (https://github.com/ethcore/parity/pull/4419)

Still TODO (Future work) -

- Rename GasPriceEditor to TxEditor (fits in closer to actual use now)
- Adjust styling of the Date/Time popups (they don't quite fit)
- Add tests for the UI components added
- Add tests for RPC formatters
- Expand tests for store transaction overrides

### CODE REVIEW

File: js/src/api/format/input.js [hunk 1]
```diff
@@ -127,6 +127,18 @@ export function inNumber16 (number) {
   return inHex(bn.toString(16));
 }
 
+export function inOptionsCondition (condition) {
+  if (condition) {
+    if (condition.block) {
+      condition.block = condition.block ? inNumber10(condition.block) : null;
+    } else if (condition.time) {
+      condition.time = inNumber10(Math.floor(condition.time.getTime() / 1000));
```
Comment (ngotchac) [inline js/src/api/format/input.js]: Is there any other RPC calls with time/date involved ? Might be useful to have an `inTime` formatter to make it DRY

File: js/src/modals/Transfer/store.js [hunk 2]
```diff
@@ -639,15 +624,12 @@ export default class TransferStore {
     const to = (isEth && !isWallet) ? this.recipient
       : (this.isWallet ? this.wallet.address : this.token.address);
 
-    const options = {
+    const options = this.gasStore.overrideTransaction({
```
Comment (ngotchac) [inline js/src/modals/Transfer/store.js]: `applyToTransaction` maybe ?

File: js/src/api/format/output.js [hunk 3]
```diff
@@ -221,6 +221,18 @@ export function outSyncing (syncing) {
   return syncing;
 }
 
+export function outTransactionCondition (condition) {
```
Comment (ngotchac) [inline js/src/api/format/output.js]: Add test ?

File: js/src/api/format/input.js [hunk 4]
```diff
@@ -127,6 +127,18 @@ export function inNumber16 (number) {
   return inHex(bn.toString(16));
 }
 
+export function inOptionsCondition (condition) {
```
Comment (ngotchac) [inline js/src/api/format/input.js]: Add test ?

File: js/src/ui/Form/InputDate/inputDate.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+import { DatePicker } from 'material-ui';
+import React, { Component, PropTypes } from 'react';
+
+import Label from '../Label';
+
+import styles from './inputDate.css';
+
+// NOTE: Has to be larger than Signer overlay Z, aligns with ../InputTime
+const DIALOG_STYLE = { zIndex: 10010 };
```
Comment (ngotchac) [inline js/src/ui/Form/InputDate/inputDate.js]: Not for this PR, but shouldn't we have some kind of central file to list all the zIndexes ? Would be easier to get a grasp of what's on top of what

File: js/src/ui/Form/InputDate/inputDate.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+  };
+
+  render () {
+    const { className, hint, label, onChange, value } = this.props;
+
+    return (
+      <div className={ [styles.container, className].join(' ') }>
+        <Label label={ label } />
+        <DatePicker
+          autoOk
```
Comment (ngotchac) [inline js/src/ui/Form/InputDate/inputDate.js]: Best prop name ! :+1:

File: js/src/ui/MethodDecoding/methodDecoding.js [hunk 7]
```diff
@@ -128,15 +129,25 @@ class MethodDecoding extends Component {
...
+    if (!condition) {
       return null;
     }
 
-    return (
-      <span>, { historic ? 'Submitted' : 'Submission' } at block <span className={ styles.highlight }>#{ minBlock.toFormat(0) }</span></span>
-    );
+    if (condition.block && condition.block.gt(0)) {
+      return (
+        <span>, { historic ? 'Submitted' : 'Submission' } at block <span className={ styles.highlight }>#{ condition.block.toFormat(0) }</span></span>
```
Comment (ngotchac) [inline js/src/ui/MethodDecoding/methodDecoding.js]: Use `FormattedMessage` ?

File: js/src/ui/MethodDecoding/methodDecoding.js [hunk 8]
```diff
@@ -128,15 +129,25 @@ class MethodDecoding extends Component {
...
+      );
+    }
+
+    if (condition.time) {
+      return (
+        <span>, { historic ? 'Submitted' : 'Submission' } at <span className={ styles.highlight }>{ moment(condition.time).format('LLLL') }</span></span>
```
Comment (ngotchac) [inline js/src/ui/MethodDecoding/methodDecoding.js]: Use `FormattedMessage` ?

File: js/src/views/Signer/components/TransactionMainDetails/transactionMainDetails.js [hunk 9]
```diff
@@ -91,7 +91,7 @@ export default class TransactionMainDetails extends Component {
       <div className={ styles.editButtonRow }>
         <Button
           icon={ <MapsLocalGasStation /> }
-          label='Edit gas/gasPrice'
+          label='Edit conditions/gas/gasPrice'
```
Comment (ngotchac) [inline js/src/views/Signer/components/TransactionMainDetails/transactionMainDetails.js]: Use `FormattedMessage` ?

File: js/src/api/format/input.js (hunk 4 above)
Comment (jacogr) [inline js/src/api/format/input.js]: Honestly don't have the space to add it here for release. Hence the comment in the PR - I need tests for all of these.

Comment (ngotchac) [inline js/src/api/format/input.js]: Fair enough ! :+1:

File: js/src/modals/Transfer/store.js (hunk 2 above)
Comment (jacogr) [inline js/src/modals/Transfer/store.js]: I didn't want to change the naming anywhere, so kept it as-is. If it bothers, can do it once there is no pressure.

File: js/src/api/format/output.js (hunk 3 above)
Comment (jacogr) [inline js/src/api/format/output.js]: As above. No tests, not in the timeframe allowed.

File: js/src/ui/MethodDecoding/methodDecoding.js (hunk 7 above)
Comment (jacogr) [inline js/src/ui/MethodDecoding/methodDecoding.js]: Historic, didn't want to change - once I do I need to do the whole file, tried to keep changes to the bare minimum.

File: js/src/ui/MethodDecoding/methodDecoding.js (hunk 8 above)
Comment (jacogr) [inline js/src/ui/MethodDecoding/methodDecoding.js]: As above.

File: js/src/views/Signer/components/TransactionMainDetails/transactionMainDetails.js (hunk 9 above)
Comment (jacogr) [inline js/src/views/Signer/components/TransactionMainDetails/transactionMainDetails.js]: As above.

File: js/src/ui/MethodDecoding/methodDecoding.js (hunk 7 above)
Comment (jacogr) [inline js/src/ui/MethodDecoding/methodDecoding.js]: The more that gets changed, the more that can go wrong. 

I actually have started with the MethodDecoding -> FormattedMessage - this file is arguably the worst I've seen out of all the ones I've done. It is not straight forward, will required a lot of (useful) refactoring to get right.

**Comment by ngotchac** [comment]
All of the comments have been addressed.
Some UI issues:

- The block number input is not a Number/Integer input (can't use `<up>` and `<down>` the augment/decrement)
- The block number input should have a default value of the current block number
- There should be an option to reset this block number input to the current block number

**Comment by jacogr** [comment]
Ok, going to put those in the "I'm not fixing that here" bucket.

- blockNumbers function exactly the way it has been in the previous iteration, only difference is the component they are rendered on.

- blockNumber is not available directly in current components, so would have to bring in some more sweeping changes (did briefly look at it - I believe the min block should be fixed to the current blockNumber as well, instead of allowing all the way down to 0 as it currently does)

- Changing the blockNumber input to `TypedInput` means we would need to do it for the rest of the elements (w.g. gas/gasPrice) as well. Nice enhancement, not going to take the risk here since I believe the gas/gasPrice itself has some assumptions about what they get in.

- Like the reset option idea. (Not here, enhancement territory)

Help always appreciated, all these are great to have, not critical to get the feature out and usable.

**Comment by ngotchac** [comment]
@jacogr This should work :

We should **at least** include the `Input` type as `number` with a minimum set to 1.
If think it's safe to include the rest as well.. Your call

```diff
diff --git a/js/src/modals/ExecuteContract/AdvancedStep/advancedStep.js b/js/src/modals/ExecuteContract/AdvancedStep/advancedStep.js
index 268ca1f..b7d8918 100644
--- a/js/src/modals/ExecuteContract/AdvancedStep/advancedStep.js
+++ b/js/src/modals/ExecuteContract/AdvancedStep/advancedStep.js
@@ -30,7 +30,7 @@ export default class AdvancedStep extends Component {
 
     return (
       <div className={ styles.gaseditor }>
-        <GasPriceEditor store={ gasStore } />
+        <GasPriceEditor gasStore={ gasStore } />
       </div>
     );
   }
diff --git a/js/src/modals/Transfer/Extras/extras.js b/js/src/modals/Transfer/Extras/extras.js
index 61b99b2..253b324 100644
--- a/js/src/modals/Transfer/Extras/extras.js
+++ b/js/src/modals/Transfer/Extras/extras.js
@@ -40,7 +40,7 @@ export default class Extras extends Component {
         { this.renderData() }
         <div className={ styles.gaseditor }>
           <GasPriceEditor
-            store={ gasStore }
+            gasStore={ gasStore }
             onChange={ onChange }
           />
         </div>
diff --git a/js/src/ui/GasPriceEditor/gasPriceEditor.js b/js/src/ui/GasPriceEditor/gasPriceEditor.js
index 35723bd..81ca80f 100644
--- a/js/src/ui/GasPriceEditor/gasPriceEditor.js
+++ b/js/src/ui/GasPriceEditor/gasPriceEditor.js
@@ -18,6 +18,7 @@ import BigNumber from 'bignumber.js';
 import { observer } from 'mobx-react';
 import React, { Component, PropTypes } from 'react';
 import { FormattedMessage } from 'react-intl';
+import { connect } from 'react-redux';
 
 import { Input, InputDate, InputTime, RadioButtons } from '../Form';
 import GasPriceSelector from '../GasPriceSelector';
@@ -56,23 +57,38 @@ const CONDITION_VALUES = [
 ];
 
 @observer
-export default class GasPriceEditor extends Component {
+class GasPriceEditor extends Component {
   static contextTypes = {
     api: PropTypes.object.isRequired
   };
 
   static propTypes = {
+    blockNumber: PropTypes.object.isRequired,
     children: PropTypes.node,
-    onChange: PropTypes.func,
-    store: PropTypes.object.isRequired
+    gasStore: PropTypes.object.isRequired,
+    onChange: PropTypes.func
   }
 
   static Store = Store;
 
+  componentWillMount () {
+    const { blockNumber, gasStore } = this.props;
+
+    gasStore.updateBlockNumber(blockNumber);
+  }
+
+  componentWillReceiveProps (nextProps) {
+    if (!this.props.blockNumber.equals(nextProps.blockNumber)) {
+      const { blockNumber, gasStore } = nextProps;
+
+      gasStore.updateBlockNumber(blockNumber);
+    }
+  }
+
   render () {
     const { api } = this.context;
-    const { children, store } = this.props;
-    const { conditionType, errorGas, errorPrice, errorTotal, estimated, gas, histogram, price, priceDefault, totalValue } = store;
+    const { children, gasStore } = this.props;
+    const { conditionType, errorGas, errorPrice, errorTotal, estimated, gas, histogram, price, priceDefault, totalValue } = gasStore;
 
     const eth = api.util.fromWei(totalValue).toFormat();
     const gasLabel = `gas (estimated: ${new BigNumber(estimated).toFormat()})`;
@@ -145,7 +161,8 @@ export default class GasPriceEditor extends Component {
   }
 
   renderConditions () {
-    const { conditionType, condition, conditionBlockError } = this.props.store;
+    const { blockNumber = 0 } = this.props;
+    const { conditionType, condition, conditionBlockError } = this.props.gasStore;
 
     if (conditionType === CONDITIONS.NONE) {
       return null;
@@ -170,6 +187,9 @@ export default class GasPriceEditor extends Component {
                 />
               }
               onChange={ this.onChangeConditionBlock }
+              min={ blockNumber }
+              step={ 1 }
+              type='number'
               value={ condition.block }
             />
           </div>
@@ -220,28 +240,43 @@ export default class GasPriceEditor extends Component {
   }
 
   onEditGas = (event, gas) => {
-    const { store, onChange } = this.props;
+    const { gasStore, onChange } = this.props;
 
-    store.setGas(gas);
+    gasStore.setGas(gas);
     onChange && onChange('gas', gas);
   }
 
   onEditGasPrice = (event, price) => {
-    const { store, onChange } = this.props;
+    const { gasStore, onChange } = this.props;
 
-    store.setPrice(price);
+    gasStore.setPrice(price);
     onChange && onChange('gasPrice', price);
   }
 
   onChangeConditionType = (conditionType) => {
-    this.props.store.setConditionType(conditionType.key);
+    this.props.gasStore.setConditionType(conditionType.key);
   }
 
   onChangeConditionBlock = (event, blockNumber) => {
-    this.props.store.setConditionBlockNumber(blockNumber);
+    this.props.gasStore.setConditionBlockNumber(blockNumber);
   }
 
   onChangeConditionDateTime = (event, datetime) => {
-    this.props.store.setConditionDateTime(datetime);
+    this.props.gasStore.setConditionDateTime(datetime);
   }
 }
+
+// export default GasPriceEditor;
+
+function mapStateToProps (state) {
+  const { blockNumber } = state.nodeStatus;
+
+  return {
+    blockNumber
+  };
+}
+
+export default connect(
+  mapStateToProps,
+  null
+)(GasPriceEditor);
diff --git a/js/src/ui/GasPriceEditor/gasPriceEditor.spec.js b/js/src/ui/GasPriceEditor/gasPriceEditor.spec.js
index 9a42e89..a2337a2 100644
--- a/js/src/ui/GasPriceEditor/gasPriceEditor.spec.js
+++ b/js/src/ui/GasPriceEditor/gasPriceEditor.spec.js
@@ -40,7 +40,7 @@ const store = {
 describe('ui/GasPriceEditor', () => {
   it('renders', () => {
     expect(shallow(
-      <GasPriceEditor store={ store } />,
+      <GasPriceEditor gasStore={ store } />,
       { context: { api } }
     )).to.be.ok;
   });
diff --git a/js/src/ui/GasPriceEditor/store.js b/js/src/ui/GasPriceEditor/store.js
index fb35f5d..53ce42e 100644
--- a/js/src/ui/GasPriceEditor/store.js
+++ b/js/src/ui/GasPriceEditor/store.js
@@ -80,7 +80,7 @@ export default class GasPriceEditor {
 
       switch (conditionType) {
         case CONDITIONS.BLOCK:
-          this.condition = Object.assign({}, this.condition, { block: 0 });
+          this.condition = Object.assign({}, this.condition, { block: this.blockNumber || 0 });
           break;
 
         case CONDITIONS.TIME:
@@ -201,6 +201,10 @@ export default class GasPriceEditor {
       });
   }
 
+  updateBlockNumber (blockNumber) {
+    this.blockNumber = blockNumber.toNumber();
+  }
+
   overrideTransaction = (transaction) => {
     if (this.errorGas || this.errorPrice || this.conditionBlockError) {
       return transaction;
diff --git a/js/src/views/Signer/components/TransactionPending/transactionPending.js b/js/src/views/Signer/components/TransactionPending/transactionPending.js
index a49f5c2..28d7356 100644
--- a/js/src/views/Signer/components/TransactionPending/transactionPending.js
+++ b/js/src/views/Signer/components/TransactionPending/transactionPending.js
@@ -122,7 +122,7 @@ export default class TransactionPending extends Component {
 
     return (
       <div className={ `${styles.container} ${className}` }>
-        <GasPriceEditor store={ this.gasStore }>
+        <GasPriceEditor gasStore={ this.gasStore }>
           <Button
             label='view transaction'
             onClick={ this.toggleGasEditor }
```

**Comment by jacogr** [comment]
I like it. I'm open to the `type=number` option as a stop-gap (better, non-intrusive - should be easily testable.) The rest I'm happy to merge in with master after this PR after some additional testing.

Will apply the number and see if anything breaks.

**Comment by jacogr** [comment]
Only edited on Transfer and blockNumber/gas/gasPrice seems ok.

**Comment by ngotchac** [comment]
Ok, what about this ?

```diff
diff --git a/js/src/ui/GasPriceEditor/store.js b/js/src/ui/GasPriceEditor/store.js
index 3d6931d..e7c0ae8 100644
--- a/js/src/ui/GasPriceEditor/store.js
+++ b/js/src/ui/GasPriceEditor/store.js
@@ -62,6 +62,9 @@ export default class GasPriceEditor {
 
     if (api) {
       this.loadDefaults();
+      api.eth.blockNumber().then((blockNumber) => {
+        this.blockNumber = blockNumber.toNumber();
+      });
     }
   }
 
@@ -80,7 +83,7 @@ export default class GasPriceEditor {
 
       switch (conditionType) {
         case CONDITIONS.BLOCK:
-          this.condition = Object.assign({}, this.condition, { block: 1 });
+          this.condition = Object.assign({}, this.condition, { block: this.blockNumber || 1 });
           break;
 
         case CONDITIONS.TIME:
```

**Comment by ngotchac** [comment]
(Just waiting to make sure it works ok with #4419)

**Comment by jacogr** [comment]
Merging as-is so testing with beta build can be done.

**Pull request merged**